src/main/java/org/elasticsearch/gateway/GatewayAllocator.java

```diff
@@ -62,43 +51,29 @@
 
     public static final String INDEX_RECOVERY_INITIAL_SHARDS = "index.recovery.initial_shards";
 
-    private final TransportNodesListGatewayStartedShards listGatewayStartedShards;
-
-    private final TransportNodesListShardStoreMetaData listShardStoreMetaData;
-
-    private final ConcurrentMap<ShardId, Map<DiscoveryNode, TransportNodesListShardStoreMetaData.StoreFilesMetaData>> cachedStores = ConcurrentCollections.newConcurrentMap();
-
-    private final ConcurrentMap<ShardId, ObjectLongHashMap<DiscoveryNode>> cachedShardsState = ConcurrentCollections.newConcurrentMap();
-
-    private final TimeValue listTimeout;
-
+    private final AsyncShardFetch.Started asyncFetchStarted;
+    private final AsyncShardFetch.Store asyncFetchStore;
     private final String initialShards;
 
     @Inject
-    public GatewayAllocator(Settings settings,
-                            TransportNodesListGatewayStartedShards listGatewayStartedShards, TransportNodesListShardStoreMetaData listShardStoreMetaData) {
+    public GatewayAllocator(Settings settings, AsyncShardFetch.Started asyncFetchStarted, AsyncShardFetch.Store asyncFetchStore) {
         super(settings);
-        this.listGatewayStartedShards = listGatewayStartedShards;
-        this.listShardStoreMetaData = listShardStoreMetaData;
+        this.asyncFetchStarted = asyncFetchStarted;
+        this.asyncFetchStore = asyncFetchStore;
 
-        this.listTimeout = settings.getAsTime("gateway.list_timeout", settings.getAsTime("gateway.local.list_timeout", TimeValue.timeValueSeconds(30)));
         this.initialShards = settings.get("gateway.initial_shards", settings.get("gateway.local.initial_shards", "quorum"));
 
-        logger.debug("using initial_shards [{}], list_timeout [{}]", initialShards, listTimeout);
+        logger.debug("using initial_shards [{}]", initialShards);
     }
 
     public void applyStartedShards(StartedRerouteAllocation allocation) {
-        for (ShardRouting shardRouting : allocation.startedShards()) {
-            cachedStores.remove(shardRouting.shardId());
-            cachedShardsState.remove(shardRouting.shardId());
-        }
+        asyncFetchStarted.applyStartedShards(allocation);
```

Comment:
Maybe add a comment here about what this is doing without having to jump to the source? (removing started shards from the caches)